File: packages/frontend/src/graph/diff.ts
/**
 * Differential analysis — pure functions, no I/O.
 *
 * Two diff modes:
 *
 * 1. **trace-vs-static** (`diffTraceVsStatic`): for each node in the graph,
 *    classify as `"touched"` (≥1 hit) or `"cold"` (0 hits) using runtime
 *    coverage data.
 *
 * 2. **trace-vs-trace** (`diffTraceVsTrace`): compare two node→count maps
 *    (e.g. from two trace sessions) and classify each node as `"new"`,
 *    `"gone"`, `"hotter"`, `"colder"`, or `"same"`.
 *
 * Both functions return `DiffEntry[]` sorted by severity (most actionable
 * first); consumers can display the list directly or build a `Map` for O(1)
 * lookup by node ID.
 */

import type { Graph } from "@grackle/shared-types";
import type { RuntimeCoverage } from "./runtimeCoverage";

export type DiffStatus =
  | "touched"
  | "cold"
  | "new"
  | "gone"
  | "hotter"
  | "colder"
  | "same";

export interface DiffEntry {
  nodeId: string;
  status: DiffStatus;
  /** Hit count in the baseline (session A, or current session for vs-static). */
  countA: number;
  /** Hit count in the comparison (session B); 0 for trace-vs-static. */
  countB: number;
  /** countB − countA (0 for trace-vs-static). */
  delta: number;
}

/**
 * Canonical hex colours per diff status — the single source of truth shared by
 * the graph overlay (GraphCanvas) and the panel chips (DiffPanel).
 *
 * All values are #rrggbb — never oklch/hsl/CSS-var — because Sigma 3.x
 * parseColor only accepts hex/rgb (ADR-0015).  `same` is the empty string: it
 * carries no override, so the graph falls through to the node's kind colour.
 */
export const DIFF_STATUS_COLORS: Record<DiffStatus, string> = {
  hotter: "#ef4444", // red    — regression
  new: "#22c55e", // green  — new coverage
  colder: "#3b82f6", // blue   — reduced calls
  gone: "#6b7280", // gray   — no longer called
  cold: "#f59e0b", // amber  — never called
  touched: "#10b981", // emerald — covered
  same: "", // empty = fall through to kind colour
};

/** Sort key per status — lower index = more actionable = shown first. */
const STATUS_ORDER: Record<DiffStatus, number> = {
  hotter: 0,
  new: 1,
  gone: 2,
  colder: 3,
  same: 4,
  cold: 5,
  touched: 6,
};

function byStatus(a: DiffEntry, b: DiffEntry): number {
  const sa = STATUS_ORDER[a.status] ?? 9;
  const sb = STATUS_ORDER[b.status] ?? 9;
  if (sa !== sb) return sa - sb;
  return a.nodeId < b.nodeId ? -1 : a.nodeId > b.nodeId ? 1 : 0;
}

/**
 * Classify every node in `graph` as `"touched"` or `"cold"` using the
 * session's runtime coverage data.
 *
 * Cold nodes (zero hits) sort first because they are the actionable ones —
 * potential dead code or untested paths.
 */
export function diffTraceVsStatic(
  graph: Graph,
  coverage: RuntimeCoverage
): DiffEntry[] {
  const entries: DiffEntry[] = graph.nodes.map((node) => {
    // RuntimeCoverage exposes only set membership (touched/cold/hot), not
    // per-node hit counts, so countA carries no quantitative value here — the
    // status field is the signal.  We deliberately set countA/countB/delta to
    // 0 rather than a synthetic 1/2: a non-zero count would read as a real hit
    // count to any consumer that renders it (e.g. a future vs-static table).
    // Callers wanting exact counts should use diffTraceVsTrace with a raw
    // counts map.
    const status: DiffStatus = coverage.touched.has(node.id)
      ? "touched"
      : "cold";
    return { nodeId: node.id, status, countA: 0, countB: 0, delta: 0 };
  });
  return entries.sort(byStatus);
}

/**
 * Build a `DiffEntry[]` from two node→count maps.
 *
 * The universe of nodes is the union of keys in `countsA` and `countsB`,
 * plus any IDs in `graphNodeIds` (so static-graph-only nodes appear as
 * `"same"` even if both sessions never touched them).
 *
 * @param countsA  Baseline session: node_id → hit count.
 * @param countsB  Comparison session: node_id → hit count.
 * @param graphNodeIds  Optional extra node IDs from the static graph.
 */
export function diffTraceVsTrace(
  countsA: Record<string, number>,
  countsB: Record<string, number>,
  graphNodeIds: string[] = []
): DiffEntry[] {
  const allIds = new Set<string>([
    ...Object.keys(countsA),
    ...Object.keys(countsB),
    ...graphNodeIds,
  ]);

  const entries: DiffEntry[] = [];
  for (const nodeId of allIds) {
    const ca = countsA[nodeId] ?? 0;
    const cb = countsB[nodeId] ?? 0;
    const delta = cb - ca;
    let status: DiffStatus;
    if (ca === 0 && cb > 0) {
      status = "new";
    } else if (ca > 0 && cb === 0) {
      status = "gone";
    } else if (delta > 0) {
      status = "hotter";
    } else if (delta < 0) {
      status = "colder";
    } else {
      status = "same";
    }
    entries.push({ nodeId, status, countA: ca, countB: cb, delta });
  }
  return entries.sort(byStatus);
}

/** Return `true` if any entry is classified as `"hotter"` (regression). */
export function hasRegression(entries: DiffEntry[]): boolean {
  return entries.some((e) => e.status === "hotter");
}

/**
 * Build a `Map<nodeId, DiffStatus>` from a `DiffEntry[]` for O(1) lookup
 * (e.g. used by GraphCanvas to colour nodes).
 */
export function diffToOverlay(entries: DiffEntry[]): Map<string, DiffStatus> {
  return new Map(entries.map((e) => [e.nodeId, e.status]));
}

/** Count entries per status. */
export function diffCounts(entries: DiffEntry[]): Record<DiffStatus, number> {
  const counts: Record<DiffStatus, number> = {
    hotter: 0,
    new: 0,
    gone: 0,
    colder: 0,
    same: 0,
    cold: 0,
    touched: 0,
  };
  for (const e of entries) counts[e.status] = (counts[e.status] ?? 0) + 1;
  return counts;
}

File: packages/agent/src/grackle/diff.py
"""Differential analysis between two trace sessions, or a trace vs. the static graph.

Two diff modes (both pure — no I/O, no side effects):

1. **trace-vs-static** (:func:`diff_trace_vs_static`): given the node IDs in the
   static graph and a single :class:`~grackle.python_runtime.aggregates.TraceAggregates`,
   classify each node as ``"touched"`` (≥1 hit) or ``"cold"`` (0 hits).

2. **trace-vs-trace** (:func:`diff_trace_vs_trace`): compare two
   :class:`~grackle.python_runtime.aggregates.TraceAggregates` instances (baseline A
   vs. comparison B) over a shared universe of node IDs and classify each node as:

   - ``"new"``    — present in B, absent in A (0 hits in A, ≥1 in B)
   - ``"gone"``   — present in A, absent in B (≥1 hit in A, 0 in B)
   - ``"hotter"`` — more hits in B than A (potential regression)
   - ``"colder"`` — fewer hits in B than A
   - ``"same"``   — equal hits in both (including both zero)

Callers control the file-scan lifecycle by building
:class:`~grackle.python_runtime.aggregates.TraceAggregates` objects beforehand via
:func:`~grackle.python_runtime.aggregates.build_seekable` or
:meth:`~grackle.python_runtime.aggregates.TraceAggregates.build`.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Literal, TypedDict

if TYPE_CHECKING:
    from grackle.python_runtime.aggregates import TraceAggregates

DiffStatus = Literal["touched", "cold", "new", "gone", "hotter", "colder", "same"]

# Sort key for each status — lower = shown first in output
_STATUS_ORDER: dict[str, int] = {
    "hotter": 0,  # most actionable (regression)
    "new": 1,
    "gone": 2,
    "colder": 3,
    "same": 4,
    "cold": 5,  # trace-vs-static cold nodes
    "touched": 6,
}


class DiffEntry(TypedDict):
    """Result of classifying one node in a diff."""

    node_id: str
    #: One of the :data:`DiffStatus` literals — kept ``str`` so JSON round-trips cleanly.
    status: str
    #: Hit count in the baseline session (session A, or current session for vs-static).
    count_a: int
    #: Hit count in the comparison session (session B); 0 for trace-vs-static diffs.
    count_b: int
    #: ``count_b - count_a`` (0 for trace-vs-static diffs).
    delta: int


def diff_trace_vs_static(
    node_ids: list[str],
    aggregates: TraceAggregates,
    at_index: int | None = None,
) -> list[DiffEntry]:
    """Classify every node in *node_ids* as ``"touched"`` or ``"cold"``.

    Args:
        node_ids:   All node IDs from the static graph (the universe to classify).
        aggregates: Trace aggregates for the session under analysis.
        at_index:   Upper-bound event index (exclusive).  Defaults to the full
                    session (``len(aggregates)``).

    Returns:
        One :class:`DiffEntry` per node ID, cold nodes first (most actionable),
        then touched nodes, both groups sorted by ``node_id`` for determinism.
    """
    idx = at_index if at_index is not None else len(aggregates)
    entries: list[DiffEntry] = []
    for nid in node_ids:
        count = aggregates.cumulative_heat(nid, idx)
        status: str = "touched" if count > 0 else "cold"
        entries.append(DiffEntry(node_id=nid, status=status, count_a=count, count_b=0, delta=0))
    entries.sort(key=lambda e: (_STATUS_ORDER.get(e["status"], 9), e["node_id"]))
    return entries


def diff_trace_vs_trace(
    aggregates_a: TraceAggregates,
    aggregates_b: TraceAggregates,
    node_ids: list[str] | None = None,
    at_index_a: int | None = None,
    at_index_b: int | None = None,
) -> list[DiffEntry]:
    """Classify every node across two trace sessions.

    The universe of nodes compared is the union of nodes seen in either session
    (i.e. nodes in ``aggregates_a.node_ids | aggregates_b.node_ids``), plus any
    extra IDs supplied via *node_ids* (e.g. the full static-graph node list).
    Nodes that are in *node_ids* but have 0 hits in both sessions are classified
    as ``"same"`` and appear at the bottom of the output.

    Args:
        aggregates_a:  Aggregates for the baseline session (A).
        aggregates_b:  Aggregates for the comparison session (B).
        node_ids:      Optional explicit universe (e.g. from ``grackle parse``
                       output).  When ``None``, the union of nodes observed in
                       the two sessions is used.
        at_index_a:    Upper-bound index for session A; defaults to full session.
        at_index_b:    Upper-bound index for session B; defaults to full session.

    Returns:
        One :class:`DiffEntry` per node, sorted by severity:
        ``hotter`` → ``new`` → ``gone`` → ``colder`` → ``same``.
    """
    idx_a = at_index_a if at_index_a is not None else len(aggregates_a)
    idx_b = at_index_b if at_index_b is not None else len(aggregates_b)

    all_ids: set[str] = set(aggregates_a.node_ids | aggregates_b.node_ids)
    if node_ids is not None:
        all_ids |= set(node_ids)

    entries: list[DiffEntry] = []
    for nid in all_ids:
        ca = aggregates_a.cumulative_heat(nid, idx_a)
        cb = aggregates_b.cumulative_heat(nid, idx_b)
        delta = cb - ca
        if ca == 0 and cb > 0:
            status = "new"
        elif ca > 0 and cb == 0:
            status = "gone"
        elif delta > 0:
            status = "hotter"
        elif delta < 0:
            status = "colder"
        else:
            status = "same"
        entries.append(DiffEntry(node_id=nid, status=status, count_a=ca, count_b=cb, delta=delta))

    entries.sort(key=lambda e: (_STATUS_ORDER.get(e["status"], 9), e["node_id"]))
    return entries


def has_regression(entries: list[DiffEntry]) -> bool:
    """Return ``True`` if any entry is classified as ``"hotter"`` (regression).

    Intended for CI use via ``grackle diff A.jsonl B.jsonl``; a non-zero exit
    code is appropriate when this returns ``True``.
    """
    return any(e["status"] == "hotter" for e in entries)
